File: 729-my-calendar-i/729-my-calendar-i.java
class MyCalendar {

    TreeSet<int[]> booked;

    public MyCalendar() {
        booked = new TreeSet<>((a, b) -> a[0] - b[0]);
    }

    public boolean book(int start, int end) {
        int[] meeting = new int[]{start, end};
        if(booked.isEmpty()) {
            booked.add(meeting);
            return true;
        }
        int[] tmp = new int[]{end, 0};
        int[] arr = booked.ceiling(tmp);
        int[] prev = (arr == null) ? booked.last() : booked.lower(tmp);
        if(prev == null || start >= prev[1]) {
            booked.add(meeting);
            return true;
        }
        return false;
    }

}

/**
 * Your MyCalendar object will be instantiated and called as such:
 * MyCalendar obj = new MyCalendar();
 * boolean param_1 = obj.book(start,end);
 */

File: 729-my-calendar-i/729-my-calendar-i.py
class MyCalendar:

    def __init__(self):
        self.booked = []

    def book(self, start: int, end: int) -> bool:
        i = bisect.bisect_right(self.booked, start)
        if i & 1 or (i < len(self.booked) and self.booked[i] < end):
            return False
        self.booked[i:i] = [start, end]
        return True

# Your MyCalendar object will be instantiated and called as such:
# obj = MyCalendar()
# param_1 = obj.book(start,end)